File: LeetCode-C++/LeetCode-C++/二叉树/_114_二叉树展开为链表.cpp
//
//  _114_二叉树展开为链表.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/11/25.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/flatten-binary-tree-to-linked-list/

#include <stdio.h>
#include "TreeNode.hpp"

class Solution {
public:
    void flatten(TreeNode* root) {
        if (root == nullptr) return;
        
        if (root->left != nullptr) {
            // 保留right
            TreeNode* right = root->right;
            // 将left嫁接到right
            root->right = root->left;
            // 清空left
            root->left = nullptr;
            // 将right嫁接到root的最右下角
            TreeNode* rightMost = root->right;
            while (rightMost->right != nullptr) {
                rightMost = rightMost->right;
            }
            rightMost->right = right;
        }
        flatten(root->right);
    }
};

/**
     1
    / \
   2   5
  / \   \
 3   4   6
 将其展开为：

 1
  \
   2
    \
     3
      \
       4
        \
         5
          \
           6
 */
/**
int main(int argc, const char * argv[]) {
    TreeNode* root = generateTreeByLevel({1,2,5,3,4,INT_MAX,6});
    Solution solution = Solution();
    solution.flatten(root);
    while (root) {
        printf("%d ", root->val);
        root = root->right;
    }
    printf("\n");
    return 0;
}
*/

File: LeetCode-Swift/LeetCode-Swift/二叉树/_114_二叉树展开为链表.swift
//
//  _114_二叉树展开为链表.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/11/25.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/flatten-binary-tree-to-linked-list/

import Foundation

class _114_二叉树展开为链表 {
    func flatten(_ root: TreeNode?) {
        guard root != nil else {
            return
        }
        
        if root?.left != nil {
            // 保留right
            let right = root?.right
            // 将left嫁接到right
            root?.right = root?.left
            // 清空left
            root?.left = nil
            // 将旧的right嫁接到root的最右下角
            var rightMost = root?.right
            while rightMost?.right != nil {
                rightMost = rightMost?.right
            }
            rightMost?.right = right
        }
        flatten(root?.right)
    }
}

/**
 输入：
        1
        / \
       2   5
       / \    \
      3   4    6
 输出：
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
 */
func test_114() {
    var root = Optional(TreeNode(1))
    root?.left = TreeNode(2)
    root?.right = TreeNode(5)
    root?.left?.left = TreeNode(3)
    root?.left?.right = TreeNode(4)
    root?.right?.right = TreeNode(6)
    
    let solution = _114_二叉树展开为链表()
    solution.flatten(root)
    var result = [String]()
    while root != nil {
        result.append(String(root!.val))
        root = root?.right
    }
    print(result)
}
